frontend/src/components/widgets/DataFrame/columns/utils.ts

```diff
@@ -358,3 +365,181 @@ export function formatNumber(value: number, maxPrecision = 4): string {
+      return isDateNotNaN(modifiedDate) || isDateNotNaN(new Date(date))
+    }
+    return isDateNotNaN(new Date(date))
+  } catch (error) {
+    logError(error)
+    return false
+  }
+}
+
+export function isStringButNumber(val: any): boolean {
+  if (typeof val === "string" && !Number.isNaN(Number(val))) {
+    return true
+  }
+  return false
+}
+
+export function isDateNotNaN(date: Date): boolean {
+  return !Number.isNaN(date.getTime())
+}
+
+export function removeZeroMillisecondsInISOString(date: string): string {
+  return date.replace(".000", "")
+}
+
+export function removeTInString(date: string): string {
+  return date.replace("T", " ")
+}
+
+export function getDateCell(
+  props: BaseColumnProps,
+  data: any,
+  type: PythonDateType
+): GridCell {
+  const defaultFormat = getDefaultFormatDateCell(type)
+
+  const parameters = {
+    ...(props.columnTypeMetadata || {}),
+  }
+
+  const cellTemplate = {
+    kind: GridCellKind.Custom,
+    allowOverlay: true,
+    copyData: "",
+    contentAlign: props.contentAlignment,
+    data: {
+      kind: "DatetimePickerCell",
+      date: undefined,
+      displayDate: "",
+      format: parameters.format ?? defaultFormat,
+      type,
+    },
+  } as DatetimePickerCell
+
+  if (isNullOrUndefined(data)) {
+    return {
+      ...cellTemplate,
+      allowOverlay: true,
+      // missing value
+      copyData: "",
+      isMissingValue: true,
+      data: {
+        kind: "DatetimePickerCell",
+        date: undefined,
+        displayDate: "",
+        format: cellTemplate.data.format,
+        type,
+      },
+    } as DatetimePickerCell
+  }
+
+  try {
+    // Python datetime uses microseconds, but JS & Moment uses milliseconds
+    if (typeof data === "bigint") {
+      data = Number(data) / 1000
+    }
+
+    if (!isValidDate(data)) {
+      return getErrorCell(`Incompatible Date value: ${data}`)
+    }
+
+    let dataDate: Date
+    if (isStringButNumber(data)) {
+      // 60000 => 60 minute / 1 second * 100 millisecond / 1 sec
+      dataDate = new Date(
+        Number(data) + new Date().getTimezoneOffset() * 60000
+      )
+    } else {
+      // safe to do new Date() because checked through isValidDate()
+      dataDate = new Date(data)
+    }
+
+    const copyData = getCopyDataForDate(dataDate, type)
+    const displayDate = removeTInString(
+      removeZeroMillisecondsInISOString(
+        moment.utc(dataDate).format(cellTemplate.data.format)
+      )
+    )
+    return {
+      ...cellTemplate,
+      allowOverlay: true,
+      copyData,
+      data: {
+        kind: "DatetimePickerCell",
+        date: dataDate,
+        displayDate,
+        format: cellTemplate.data.format,
+        type,
+      },
+    } as DatetimePickerCell
+  } catch (error) {
+    return getErrorCell(`Incompatible date value: ${data}`)
+  }
+}
+
+export function getDefaultFormatDateCell(type: PythonDateType): string {
+  switch (type) {
+    case "date":
+      return "YYYY / MM / DD"
```

Comment:
While this format is a bit easier to read, I would prefer to use the ISO-compatible date format: `YYYY-MM-DD` here instead for consistency reasons.